index.bs

```diff
@@ -787,14 +718,595 @@ the following list:
+</div>
+
+<hr>
+
+When [=the WebSocket closing handshake is started=] for a {{WebSocketStream}}
+|stream|, the user agent must [=queue a global task=] on the [=WebSocket task source=] given
+|stream|'s [=relevant global object=] to change the |stream|'s [=WebSocketStream/ready state=] to
+{{WebSocket/CLOSING}} if it is not already. [[!WSP]]
+
+<hr>
+
+<div algorithm="handle WebSocketStream connection close">
+
+When [=the WebSocket connection is closed=] for a {{WebSocketStream}} |stream|, possibly
+[=cleanly=], the user agent must [=queue a global task=] on the [=WebSocket task source=] given
+|stream|'s [=relevant global object=] to run the following substeps:
+
+ 1. Change the [=WebSocketStream/ready state=] to {{WebSocket/CLOSED}} (3).
+ 1. If |stream|'s [=WebSocketStream/handshake aborted=] is true, then return.
+ 1. If |stream|'s [=WebSocketStream/was ever connected=] is false, then [=reject=] |stream|'s
+    [=opened promise=] with a [=new=] WebSocketError.
+ 1. Let |code| be [=the WebSocket connection close code=].
+ 1. Let |reason| be the result of applying [=UTF-8 decode without BOM=] to [=the WebSocket
+    connection close reason=].
+ 1. If the connection was closed [=cleanly=],
+  1. [=ReadableStream/Close=] |stream|'s [=WebSocketStream/readable stream=].
+  1. [=WritableStream/Error=] |stream|'s [=WebSocketStream/writable stream=] with an
+     "{{InvalidStateError}}" {{DOMException}} indicating that a closed {{WebSocketStream}} cannot be
+     written to.
+  1. [=Resolve=] |stream|'s [=closed promise=] with {{WebSocketCloseInfo}} «[
+    "{{WebSocketCloseInfo/closeCode}}" → |code|, "{{WebSocketCloseInfo/reason}}" → |reason| ]».
+ 1. Otherwise,
+  1. Let |error| be a [=new=]
+     {{WebSocketError}} whose [=WebSocketError/closeCode=] is |code| and [=WebSocketError/reason=]
+     is |reason|.
+  1. [=ReadableStream/Error=] |stream|'s [=WebSocketStream/readable stream=] with |error|.
+  1. [=WritableStream/Error=] |stream|'s [=WebSocketStream/writable stream=] with |error|.
+  1. [=Reject=] |stream|'s [=closed promise=] with |error|.
+</div>
+
+
+
+## Stream operations ## {#stream-operations}
+
+<div algorithm="pull algorithm">
+
+To <dfn>pull bytes</dfn> from a {{WebSocketStream}} |stream|, if |stream| is currently applying
+backpressure, release backpressure.
+
+ <p class=note>If any messages are queued, one will be handled immediately as a result.
+</div>
+
+<hr>
+
+<div algorithm="cancel algorithm">
+
+To <dfn>cancel</dfn> a {{WebSocketStream}} |stream| given |reason|, [=close using reason=] giving
+|stream| and |reason|.
+</div>
+
+<hr>
+
+<div algorithm="write algorithm">
+
+To <dfn>write</dfn> |chunk| to a {{WebSocketStream}} |stream|:
+ 1. Let |promise| be [=a new promise=] created in |stream|'s [=relevant realm=].
+ 1. If |chunk| is a {{BufferSource}},
+  1. Let |data| be a [=get a copy of the buffer source|copy of the bytes=] given |chunk|.
+  1. Let |opcode| be a binary frame opcode.
```

Comment:
We tend to think of variables as block-scoped, so it'd be best to declare data and opcode outside the if statements (e.g. let them both be null.)